content/Stop creating services for everything - Knowing YAAF & the FormObject Pattern.md

```diff
@@ -0,0 +1,143 @@
+  end
+end
+```
+
+It looks terrible, doesn't it? Maybe our first thinking about that piece of code is to make a refactoring moving the creation of the post to a service named `PostCreationService`, which could be useful and it might be used in the future in another part of the system. But... what have we said about _reinventing the wheel_?
+
+Using **YAAF** we should create a new `PostForm` class that is going to encapsulate all the logic of post creation and related models inside it. And is very simple to implement it! Look at this code:
+
+```ruby
+# app/forms/registration_form.rb
+
+class PostForm < ApplicationForm
+  attr_accessor :post_form_params
+  validate :amount_of_tags
+
+  def initialize(args = {})
+    @post_form_params = args
+    @models = [post, category, tags].flatten.compact
+  end
+
+  def post
+    @post ||= Post.new(post_form_params[:post]).tap do |post|
+      post.category = category
+      post.tags = tags
+    end
+  end
+
+  def category
+    return [] if post_form_params[:category_name].blank?
+
+    @category ||= Category.find_or_initialize_by(name: post_form_params[:category_name])
+  end
+
+  def tags
+    return [] if post_form_params[:tags].blank?
+
+    @tags ||= post_form_params[:tags].map do |tag|
+      tag[:id].present? ? Tag.find(tag[:id]) : Tag.find_or_initialize_by(name: tag[:name])
+    end
+  end
+
+  private
+
+  def amount_of_tags
+    return if tags.size.between?(1, 3)
+
+    errors.add(:base, "You can't assign more than three tags to a post")
+  end
+end
+```
+
+**Note:** We have also added a custom validation named `amount_of_tags`, YAAF helps us to encapsulate business rules in our Form Object.
+
+And then in our controller, we have the following:
+
+```ruby
+class Api::V1::PostsController < Api::V1::ApiController
+  def index
+    @posts = Post.all
+  end
+
+  def create
+    form = PostForm.new(post_form_params)
+    form.save!
+    @post = form.post
+  end
+
+  private
+
+  def post_form_params
+    params.permit(:category_name, tags: %i[id name], post: %i[title body publisher_id category_id])
+  end
+end
+```
+
+**Tip:** Having an `ApplicationForm` which inherits from `YAAF::Form` is a good practice.
+
+That's it, now we have a `PostForm` which encapsulates all the persistency logic of post/tags/categories, leaving our controller and models clean and the code is easy to follow.
+
+Another good thing is that **YAAF** provides a similar API to `ActiveModel` models so you can treat them interchangeably.
+
+## Why not a Service or PORO's?
+
+- **Make customized Services or PORO's could be disorganized** if you're working in a team.
+- YAAF helps you to apply the **Form Pattern** in an easy way.
+- YAAF is only **64 lines long**.
+- It's **well tested and maintained**.
+- It helps you keep your models, views and controllers thin by providing a better place where to put business logic. In the end, this will **improve the quality of your codebase and make it easier to maintain and extend**.
```

Comment:
```suggestion
- It helps you keep your models, views, and controllers thin by providing a better place to put business logic. In the end, this will **improve the quality of your codebase and make it easier to maintain and extend**.
```